ocaml/xenopsd/lib/xenops_utils.ml

```diff
@@ -115,27 +115,25 @@ let ignore_bool (_ : bool) = ()
 let ignore_int (_ : int) = ()
 
 (* Recode an incoming string as valid UTF-8 *)
+let utf8_add_if_valid buf _ next =
+  let uchar = match next with
+  | `Malformed _ -> Uutf.u_rep
+  | `Uchar u -> u in
+  Uutf.Buffer.add_utf_8 buf uchar;
+  buf
+
+let is_valid prev _ = function
```

Comment:
It is somewhat unusual to have the predicate take a second argument but here we might want to do it for efficiency. `is_valid_and` might convey this slightly better.